pkg/api/server.go

```diff
@@ -61,28 +73,59 @@ func authenticateHandler() func(http.Handler) http.Handler {
 				return
 			}
 
-			c := config.GetInstance()
 			ctx := r.Context()
 
-			// handle redirect if no user and user is required
-			if userID == "" && c.HasCredentials() && !allowUnauthenticated(r) {
-				// if we don't have a userID, then redirect
-				// if graphql was requested, we just return a forbidden error
-				if r.URL.Path == "/graphql" {
-					w.Header().Add("WWW-Authenticate", `FormBased`)
-					w.WriteHeader(http.StatusUnauthorized)
+			if c.HasCredentials() {
+				// authentication is required
+				if userID == "" && !allowUnauthenticated(r) {
+					//authentication was not received, redirect
+					// if graphql was requested, we just return a forbidden error
+					if r.URL.Path == "/graphql" {
+						w.Header().Add("WWW-Authenticate", `FormBased`)
+						w.WriteHeader(http.StatusUnauthorized)
+						return
+					}
+
+					// otherwise redirect to the login page
+					u := url.URL{
+						Path: "/login",
+					}
+					q := u.Query()
+					q.Set(returnURLParam, r.URL.Path)
+					u.RawQuery = q.Encode()
+					http.Redirect(w, r, u.String(), http.StatusFound)
 					return
 				}
-
-				// otherwise redirect to the login page
-				u := url.URL{
-					Path: "/login",
+			} else {
+				//authentication is not required
+				//security fix: traffic from the public internet with no auth is disallowed
+				if !c.GetDangerousAllowPublicWithoutAuth() && !c.IsNewSystem() {
+					requestIPString := r.Header.Get("X-FORWARDED-FOR")
```

Comment:
Using `X-FORWARDED-FOR` without any checks doesn't make it secure. An attacker can just include the `X-FORWARDED-FOR` header in the request with a private IP address in it and the request will be accepted, right?

This needs a "trusted proxies" settings so the remote IP is validated to be a proxy which may be used, in that case `X-FORWARDED-FOR` may be accepted, otherwise the request should still be denied (/trip the wire). The "trusted proxies" could then still be set to a default of the private IP ranges for example.

Reply:
Personally I would most likely implement it as:
1. Check if `X-FORWARDED-FOR` is set
2. If it's set validate whether the remote IP of the request complies with a "trusted proxies" setting
3. If the remote IP doesn't comply deny the request (but don't trip the wire)
4. If `X-FORWARDED-FOR` is set it can now safely be used

This would require users of a proxy to set up this trusted proxies setting, but IMO that should be fine and at least makes it secure by default.
If that's not desirable I would allow private subnets by default until the user sets it.

Reply:
Another comment about `X-FORWARDED-FOR`. As far as I know this header can contain multiple addresses. This in case that a request passes through multiple proxies. But I'm unsure what it's value is in that case (comma separated? colon separated?). But I guess that all IP addresses should be validated in that case. All except the last (or first?) must be validated as trusted proxy, and the remaining should in that case thus comply with a private subnet.

Reply:
Good catch!

Although now that I'm thinking about it, I'm tempted to just remove the header check. It's a de-facto non-standard that can contain anything and everything, so it would be very hard to validate all possible values. Also, what would the usecase be where you are accessing stash from a proxy outside its local network? If you are, you probably know enough to not expose it to the internet or to set up your own auth if you are.

Reply:
Doesn't removing it circumvent this fix?

If I'd be running a reverse proxy in my home, which does expose Stash (for example putting Traefik in front of it and using a Lets encrypt certificate for it) then Stash could be accessible from the internet again. Because the proxies IP address would be private and the IP of the actual requestor isn't validated anymore.

Reply:
I think that if you know how to set up traefik or even know what a reverse proxy is, you aren't the target of this patch. This is for the 50 or so (mostly windows gaming machines) that have all of their ports forwarded. There are some docker installs publicly accessible too, but still.

Reply:
I do get your point, but at the same time the same IMO applies to those Windows users which just add a port forward for all available ports.

In both cases an ignorant user applies a faulty configuration which makes them vulnerable. Whether it's a user which just applies a port forward for all ports or whether it's a user which has either Traefik running with the configuration to expose all containers it does find, the user manually exposing it to Traefik and having to make it publicly accessible for the HTTP ACME challenge and forgetting to set up authentication or whatever. (And there most likely are much more scenarios, using a different reverse proxy which might always expose all services it finds. The computer not being behind a NAT-router so exposing all ports to the internet as well, running Stash on a server / VPS, ...).

In both cases it is because an ignorant user applies the "wrong" configuration which exposes their Stash instance. And only protecting a small group of users thus feels "wrong" IMHO.